### PR TITLE
Fixed issue #15501

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Backport of https://github.com/arangodb/arangodb/pull/15565

* Fixed issue #15501: Regression when using "exclusive" query option?
  This fixes a regression in AQL query execution when exclusive locks are used
  for a query and the query also uses the DOCUMENT() AQL function. In this
  case, when there were more concurrent requests to the underlying collection
  than available scheduler threads for the low priority queue, the query
  start successfully acquired the exclusive lock could not get its follow-up
  requests through and starve while holding the exclusive lock.

- [x] :hankey: Bugfix
- [ ] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [ ] **integration tests**
  - [ ] **resilience tests**
- [x] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)

#### Related Information

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [x] GitHub issue / Jira ticket: https://github.com/arangodb/arangodb/issues/15501
- [ ] Design document:

### CODE REVIEW

File: arangod/Aql/RestAqlHandler.cpp [hunk 1]
```diff
@@ -873,8 +873,24 @@ RestStatus RestAqlHandler::handleFinishQuery(std::string const& idString) {
...
+  } 
+
+  if (ServerState::instance()->isDBServer()) {
+    std::vector<std::string> const& suffixes = _request->suffixes();
+
+    if (suffixes.size() == 2 && suffixes[0] == "finish") {
+      // AQL shutdown requests should have medium priority, so it can release
+      // locks etc. and unblock other pending requests
+      TRI_ASSERT(PriorityRequestLane(RequestLane::CONTINUATION) == RequestPriority::MED);
+      return RequestLane::CONTINUATION;
```
Comment (goedderz) [inline arangod/Aql/RestAqlHandler.cpp:889]: `CONTINUATION` doesn't seem to be fitting. Maybe introduce another RequestLane?

File: arangod/Aql/RestAqlHandler.cpp [hunk 2]
```diff
@@ -873,8 +873,24 @@ RestStatus RestAqlHandler::handleFinishQuery(std::string const& idString) {
 
 RequestLane RestAqlHandler::lane() const {
   if (ServerState::instance()->isCoordinator()) {
+    // continuation requests on coordinators will get medium priority,
+    // so that they don't block query parts elsewhere
+    TRI_ASSERT(PriorityRequestLane(RequestLane::CLUSTER_AQL_INTERNAL_COORDINATOR) == RequestPriority::MED);
```
Comment (goedderz) [inline arangod/Aql/RestAqlHandler.cpp:878]: If you make `PriorityRequestLane` `constexpr`, this assertion can be changed into a `static_assert`. Same for the other asserts below.

**Review by mpoeter — APPROVED** [review]

File: CHANGELOG [hunk 3]
```diff
@@ -1,6 +1,14 @@
 v3.8.5 (XXXX-XX-XX)
 -------------------
 
+* Fixed issue #15501: Regression when using "exclusive" query option?
+  This fixes a regression in AQL query execution when exclusive locks are used
+  for a query and the query also uses the DOCUMENT() AQL function. In this
+  case, when there were more concurrent requests to the underlying collection
+  than available scheduler threads for the low priority queue, the query
+  start successfully acquired the exclusive lock could not get its follow-up
+  requests through and starve while holding the exclusive lock.
```
[the remark this answers is not in the document]
Comment (jsteemann) [reply CHANGELOG:10]: Changelog adjusted, so hopefully addressed.

File: arangod/RestHandler/RestVocbaseBaseHandler.cpp [hunk 4]
```diff
@@ -635,13 +635,23 @@ std::unique_ptr<transaction::Methods> RestVocbaseBaseHandler::createTransaction(
...
     trx = std::make_unique<transaction::Methods>(std::move(ctx), collectionName, type);
   } else {
     trx = std::make_unique<transaction::Methods>(std::move(ctx));
+    if (isSideUser) {
+      // this is a call from the DOCUMENT() AQL function into an existing AQL query.
+      // locks are already acquired by the AQL transaction.
+      TRI_ASSERT(type == AccessMode::Type::READ);
+      if (type != AccessMode::Type::READ) {
+        THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_INTERNAL, "invalid access mode for request");
+      }
```
[the remark this answers is not in the document]
Comment (jsteemann) [reply arangod/RestHandler/RestVocbaseBaseHandler.cpp:653]: Addressed.

File: arangod/Aql/RestAqlHandler.cpp (hunk 2 above)
Comment (jsteemann) [reply arangod/Aql/RestAqlHandler.cpp:878]: Done

**Review by goedderz — APPROVED** [review]

File: arangod/GeneralServer/RequestLane.h [hunk 5]
```diff
@@ -166,9 +183,11 @@ inline RequestPriority PriorityRequestLane(RequestLane lane) {
...
     case RequestLane::CONTINUATION:
       return RequestPriority::MED;
-    case RequestLane::UNDEFINED: {
-      TRI_ASSERT(false);
-    }
+    case RequestLane::UNDEFINED:
+      // assume low priority for UNDEFINED. we should never get
+      // here under normal circumstances. if we do, returning the
+      // default shouldn't do any harm.
+      return RequestPriority::LOW;
```
Comment (mpoeter) [inline arangod/GeneralServer/RequestLane.h:190]: Since we are never supposed to get here, perhaps we should leave the assert in?

**Review by mpoeter — APPROVED** [review]

**Pull request merged**